crypto/src/external_commit.rs

```diff
@@ -0,0 +1,243 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+
+use core_crypto_keystore::CryptoKeystoreMls;
+use openmls::{
+    group::{GroupId, MlsGroup},
+    prelude::{MlsMessageOut, VerifiablePublicGroupState},
+};
+use openmls_traits::OpenMlsCryptoProvider;
+
+use crate::{
+    prelude::{MlsConversation, MlsConversationConfiguration},
+    CryptoError, CryptoResult, MlsCentral, MlsError,
+};
+
+impl MlsCentral {
+    pub async fn join_by_external_commit(
+        &self,
+        group_state: VerifiablePublicGroupState,
+    ) -> CryptoResult<(GroupId, MlsMessageOut)> {
+        let credentials = self.mls_client.credentials();
+        let (mut group, message) = MlsGroup::join_by_external_commit(
+            &self.mls_backend,
+            None,
+            group_state,
+            &MlsConversationConfiguration::default().as_openmls_default_configuration(),
+            &[],
+            credentials,
+        )
+        .await
+        .map_err(MlsError::from)
+        .map_err(CryptoError::from)?;
+
+        let mut buf = vec![];
+        group.save(&mut buf)?;
+        self.mls_backend
+            .key_store()
+            .mls_pending_groups_save(group.group_id().as_slice(), &buf)
+            .await
+            .map_err(CryptoError::from)?;
+        Ok((group.group_id().clone(), message))
+    }
+
+    pub async fn merge_pending_group_from_external_commit(
+        &mut self,
+        group_id: &[u8],
+        configuration: MlsConversationConfiguration,
+    ) -> CryptoResult<()> {
+        let buf = self
+            .mls_backend
+            .key_store()
+            .mls_pending_groups_load(group_id)
+            .await
+            .map_err(CryptoError::from)?;
+        let mut mls_group = MlsGroup::load(&mut &buf[..])?;
+        mls_group.merge_pending_commit().map_err(MlsError::from)?;
+        self.mls_backend
+            .key_store()
+            .mls_pending_groups_delete(group_id)
+            .await
+            .map_err(CryptoError::from)?;
+        let conversation = MlsConversation::from_mls_group(mls_group, configuration, &self.mls_backend).await?;
+        self.mls_groups.insert(group_id.to_owned(), conversation);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use mls_crypto_provider::MlsCryptoProvider;
+
+    use crate::{credential::CredentialSupplier, member::ConversationMember};
+
+    wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+    pub mod external_join {
```

Comment:
Other tests in mind:
* What if `merge_pending_group_from_external_commit` while there's no pending group ?
* What happens if PublicGroupState is not the one of the group I am joining
* If `merge_pending_group_from_external_commit` fails, can I retry it (useful from a client POV)

Also in your tests, if you can, verify that at the end pending_groups are pruned from keystore